tests/virt_autotest/validate_host_and_guests.pm

```diff
@@ -0,0 +1,68 @@
+# SUSE's openQA tests
+#
+# Copyright 2012-2022 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+#
+# Summary: Do a basic examination to the host and guests after tests run
+# Maintainer: qe-virt@suse.de
+
+package validate_host_and_guests;
+
+use strict;
+use warnings;
+use base "virt_feature_test_base";
+use testapi;
+use ipmi_backend_utils qw(reconnect_when_ssh_console_broken);
+use virt_autotest::common;
+use virt_autotest::utils qw(start_guests check_port_state check_host_health check_guest_health);
+
+sub run_test {
+
+    my %failures;
+    my $unhealthy_guests;
+    my $living_guests;
+    my $fail = check_host_health;
+    $failures{host} = $fail unless $fail eq "";
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        script_run("virsh start $guest");
+        if (script_retry("nmap $guest -PN -p ssh | grep open", delay => 2, retry => 60, die => 0) == 0) {
+            $living_guests .= $guest;
+            $fail = check_guest_health($guest);
+            unless ($fail eq "") {
+                $unhealthy_guests .= $guest;
+                $failures{$guest} = $fail;
+            }
+        }
+        else {
+            $failures{$guest} = "Not reacheable";
+            record_info("$guest: $failures{$guest}", result => 'fail');
+        }
+    }
+    if (get_var('FORCE_UPLOAD_LOGS')) {
+        virt_utils::collect_host_and_guest_logs($living_guests);
+    }
+    else {
+        virt_utils::collect_host_and_guest_logs($unhealthy_guests) unless %failures == 0;
+    }
+    die "Host or guests are not healthy!" unless %failures == 0;
+}
+
+sub prepare_run_test {
+    reconnect_when_ssh_console_broken unless defined(script_run("rm -f /root/{commands_history,commands_failure}", die_on_timeout => 0));
+    script_run("history -c");
+}
+
+sub post_fail_hook {
+    my $self = shift;
+    upload_logs("/var/log/clean_up_virt_logs.log");
+    upload_logs("/var/log/guest_console_monitor.log");
+    script_run("rm -f -r /var/log/clean_up_virt_logs.log /var/log/guest_console_monitor.log");
+    $self->upload_coredumps;
+    save_screenshot;
+}
+
+sub test_flags {
+    return {fatal => 1};
+}
+
+1;
```

Comment:
1. Rename the module to validate_system_health or other more specific names
2. Your check health functions should return explicitly, for example, 'pass' or 'fail'. Not only for good practice, but can also be called and used more efficiently.
3. Your new module should be improved a little bit, there are too many unnecessary steps, or example:
```
my %healthy_status = ();
@healthy_status{(keys %virt_autotest::common::guests)} = ();

sub prepare_run_test {
    reconnect_when_ssh_console_broken unless defined(script_run("rm -f /root/{commands_history,commands_failure}", die_on_timeout => 0));
    script_run("history -c");
}

sub run_test {
    foreach (keys %healthy_status) {
        script_run("virsh start $_");
        if (script_retry("nmap $_ -PN -p ssh | grep open", delay => 2, retry => 60, die => 0) == 0) {
            $healthy_status{$_} = check_guest_health($_);
        }
        record_info("$_: $healthy_status{$_}", result => "$healthy_status{$_}");
    }
    $healthy_status{host} = check_host_health;
    die "Host or guests are not healthy!" if (grep 'fail', values %healthy_status);
}


sub post_fail_hook {
    my $self = shift;
        if (get_var('FORCE_UPLOAD_LOGS')) {
        virt_utils::collect_host_and_guest_logs(join(' ', grep {$_ ne 'host'} keys %healthy_status));
    }
    else {
        virt_utils::collect_host_and_guest_logs(join(' ', grep { $healthy_status{$_} eq 'fail' } keys %healthy_status));
    }
    upload_logs("/var/log/clean_up_virt_logs.log");
    upload_logs("/var/log/guest_console_monitor.log");
    script_run("rm -f -r /var/log/clean_up_virt_logs.log /var/log/guest_console_monitor.log");
    $self->upload_coredumps;
    save_screenshot;
}

sub test_flags {
    return {fatal => 1};
}

1;
```

Reply:
> 1. Rename the module to validate_system_health or other more specific names

No probem.

> 2. Your check health functions should return explicitly, for example, 'pass' or 'fail'. Not only for good practice, but can also be called and used more efficiently.

Let me try.

> 3. Your new module should be improved a little bit, there are too many unnecessary steps, for example:

The hash is ok, but

- do we still collect_host_and_guest_logs for guests without ssh connection?
- FORCE_UPLOAD_LOGS is to upload logs forcefully even with a successful run, because logs might be requested in bugs or for some debugging purpose.  we cannot put it in post_fail_hook, or it will not be called if test passes.